tests/Unit/ApparentHorizons/TestHorizonManagerHelper_ElementComponent.hpp

```diff
@@ -0,0 +1,125 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <tuple>
+#include <utility>
+
+#include "ApparentHorizons/HorizonManagerComponent.hpp"
+#include "ApparentHorizons/HorizonManagerComponentActions.hpp"
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/CreateInitialElement.hpp"
+#include "Domain/Domain.hpp"
+#include "Domain/DomainCreators/DomainCreator.hpp"
+#include "Domain/Element.hpp"
+#include "Domain/ElementId.hpp"
+#include "Domain/ElementIndex.hpp"
+#include "Domain/ElementMap.hpp"
+#include "Domain/InitialElementIds.hpp"
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Parallel/Algorithm.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Printf.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp"
+#include "PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp"
+#include "PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp"
+#include "PointwiseFunctions/GeneralRelativity/GrTags.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/ApparentHorizons/TestHorizonManagerHelper_ElementActions.hpp"
+
+#include <iostream>
+
+template <class Metavariables>
+struct DgElementArray {
+  using chare_type = Parallel::Algorithms::Array;
+  using const_global_cache_tag_list = tmpl::list<>;
+  using metavariables = Metavariables;
+  using action_list = tmpl::list<>;
+  using array_index = ElementIndex<3>;
+  using initial_databox = db::compute_databox_type<
+      typename Actions::DgElementArray::InitializeElement::return_tag_list>;
+  using options = tmpl::list<typename Metavariables::domain_creator_tag>;
+  static void initialize(
+      Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache,
+      std::unique_ptr<DomainCreator<3, Frame::Inertial>>
+          domain_creator) noexcept;
+  static void execute_next_phase(
+      const typename Metavariables::Phase next_phase,
+      const Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache) {
+    if (next_phase == Metavariables::Phase::InitialCommunication) {
+      auto& dg_element_array = Parallel::get_parallel_component<DgElementArray>(
+          *(global_cache.ckLocalBranch()));
+
+      Parallel::simple_action<Actions::DgElementArray::SendNumElements<
+          HorizonManagerComponent<Metavariables>>>(dg_element_array);
```

Comment:
Can this be part of the initialization? For the linear solver I do a reduction across elements in the initialization phase so I think it's possible.

Reply:
My current understanding is that the initialization function should not depend on having other parallel components already initialized, since each component is initialized in a random order.  In this case, SendNumElements calls an Action of HorizonManagerComponent, which needs to be initialized before it is called. If SendNumElements were called during the initialization, it might call an Action on an uninitialized HorizonManagerComponent.  Those that understand actions better than me should correct me if I am wrong.